File: julia/04_array_member.jl
# One ensemble member, selected by the Slurm array task ID.
#
# The workhorse pattern for this group: independent members, one task each,
# scheduled and retried individually.

using Random, JSON3

member = parse(Int, get(ENV, "SLURM_ARRAY_TASK_ID", "1"))
outdir = get(ENV, "OUTDIR", "output")

# Seed from the member index so the run is reproducible and members differ.
rng = MersenneTwister(member)
result = (member = member, mean = sum(randn(rng, 1000)) / 1000)

mkpath(outdir)
out = joinpath(outdir, "member_" * lpad(member, 5, '0') * ".json")
open(out, "w") do io
    JSON3.write(io, result)
end
println("wrote ", out)

File: python/04_array_member.py
"""One ensemble member, selected by the Slurm array task ID.

This is the workhorse pattern for this group: independent members, one task
each, scheduled and retried individually.
"""
import argparse
import json
import os
import pathlib

import numpy as np

parser = argparse.ArgumentParser()
parser.add_argument("--member", type=int, default=None)
parser.add_argument("--outdir", type=pathlib.Path, default=pathlib.Path("output"))
args = parser.parse_args()

member = args.member
if member is None:
    member = int(os.environ["SLURM_ARRAY_TASK_ID"])

# Seed from the member index so the run is reproducible and every member differs.
rng = np.random.default_rng(member)
result = {"member": member, "mean": float(rng.normal(size=1000).mean())}

args.outdir.mkdir(parents=True, exist_ok=True)
out = args.outdir / f"member_{member:05d}.json"
out.write_text(json.dumps(result))
print(f"wrote {out}")
